pkg/tbtc/signing.go

```diff
@@ -233,3 +324,172 @@ func (srl *signingRetryLoop) qualifiedOperatorsSet() (
 
 	return chain.Addresses(qualifiedOperators).Set(), nil
 }
+
+// excludedMembersIndexes returns a list of excluded members' indexes for
+// the given qualified operators set.
+func (srl *signingRetryLoop) excludedMembersIndexes(
+	qualifiedOperatorsSet map[chain.Address]bool,
+	readyMembersIndexes []group.MemberIndex,
+) []group.MemberIndex {
+	includedMembersIndexes := make([]group.MemberIndex, 0)
+	excludedMembersIndexes := make([]group.MemberIndex, 0)
+	for i, operator := range srl.signingGroupOperators {
+		memberIndex := group.MemberIndex(i + 1)
+
+		if qualifiedOperatorsSet[operator] &&
+			slices.Contains(readyMembersIndexes, memberIndex) {
```

Comment:
I am confused. In `qualifiedOperatorsSet` we pass only `readySigningGroupOperators` to `retry.EvaluateRetryParticipantsForSigning`. From what I understand, it should not return anyone who is not on that list. Why do we check here again then?

Reply:
Yeah, this is confusing but that's the only option to make it work as expected for cases where some members of the same operator are ready and others are not. 

Suppose we have operator `O` controlling members `1`, `3`, and `5` where `1` did not announce their readiness. The `retry.EvaluateRetryParticipantsForSigning` does not care about the exact mapping between the operator and the exact members it controls. It only operates on the seat count which is `2` for operator `O` as we filtered out ready members using the `readyMembersIndexes` argument at the beginning of `qualifiedOperatorsSet`. Once we return from `qualifiedOperatorsSet` we only know a set of qualified operators (selected by the random algorithm) whose total seat count is enough to maintain the threshold. To properly take ready members (`3` and `5`) for operator `O` in the part you mentioned, we need to check against `readyMembersIndexes` again to properly take the two ready members controlled by operator `O` which are `3` and `5`. Otherwise, we would end up with `1` and `3` which is wrong as `1` did not announce their readiness.

Reply:
Instead of calling `srl.qualifiedOperatorsSet(readyMembersIndexes)` and `srl.excludedMembersIndexes` from the `start` function, can we introduce another function that would wrap `qualifiedOperatorsSet` and `excludedMembersIndexes` and would have the explanation from your comment?

Also, I wonder if we should check `slices.Contains(readyMembersIndexes, memberIndex)` as a part of this loop, or maybe have a separate one, after `qualifiedOperatorsSet` that would do it. It's more work but I think it would be easier to understand. 

We could also add another function to `retry` that would call `EvaluateRetryParticipantsForSigning` internally and did care about the seats.

Reply:
> Instead of calling srl.qualifiedOperatorsSet(readyMembersIndexes) and srl.excludedMembersIndexes from the start function, can we introduce another function that would wrap qualifiedOperatorsSet and excludedMembersIndexes and would have the explanation from your comment?

See 26b343a7.

> Also, I wonder if we should check slices.Contains(readyMembersIndexes, memberIndex) as a part of this loop, or maybe have a separate one, after qualifiedOperatorsSet that would do it. It's more work but I think it would be easier to understand.

I think that should be clear now given the change from 26b343a7

> We could also add another function to retry that would call EvaluateRetryParticipantsForSigning internally and do care about the seats.

I'd rather avoid changing anything in the `retry` package unless absolutely needed. The current interface is generic enough and it does what it is supposed to do. Also, we have consistent functions for both signing and DKG. I think all the nuances should be handled in the client packages.